Tests/SourceKitLSPTests/TargetVisibilityChangeTests.swift

```diff
@@ -0,0 +1,129 @@
+import TSCBasic
+import XCTest
+
+final class IndexVisibilityChangeTests: XCTestCase {
+
+  /// Connection and lifetime management for the service.
+  var testServer: TestSourceKitServer! = nil
+
+  /// The primary interface to make requests to the SourceKitServer.
+  var sk: TestClient! = nil
+
+  /// The server's workspace data. Accessing this is unsafe if the server does so concurrently.
+  var workspace: Workspace! = nil
+
+  /// The build system that we use to verify SourceKitServer behavior.
+  var buildSystem: TestBuildSystem! = nil
+
+  override func setUp() {
+    testServer = TestSourceKitServer()
+    buildSystem = TestBuildSystem()
+
+    let server = testServer.server!
+    workspace = Workspace(
+      rootUri: nil,
+      clientCapabilities: ClientCapabilities(),
+      toolchainRegistry: ToolchainRegistry.shared,
+      buildSetup: TestSourceKitServer.serverOptions.buildSetup,
+      underlyingBuildSystem: buildSystem,
+      index: nil,
+      indexDelegate: nil,
+      explicitIndexMode: true)
+
+    server.workspace = workspace
+    workspace.buildSystemManager.delegate = server
+
+    sk = testServer.client
+    _ = try! sk.sendSync(InitializeRequest(
+        processId: nil,
+        rootPath: nil,
+        rootURI: nil,
+        initializationOptions: nil,
+        capabilities: ClientCapabilities(workspace: nil, textDocument: nil),
+        trace: .off,
+        workspaceFolders: nil))
+  }
+
+  override func tearDown() {
+    buildSystem = nil
+    workspace = nil
+    sk = nil
+    testServer = nil
+  }
+
+  func testSchemeChanged() {
+    let targetGraph = mockBuildTargetGraph(targetStringRepr: [
+         "target://a:a": [],
+         "target://b:b": ["target://a:a"],
+         "target://c:c": ["target://b:b"],
+         "target://d:d": [],
+       ])
+
+    buildSystem.targets = targetGraph
+
+    let targetOutputResponse: (([BuildTargetIdentifier]) -> [OutputsItem]) = { (targets: [BuildTargetIdentifier]) in
+      return targets.compactMap { (targetIdentifier: BuildTargetIdentifier) in
+        let outputFile = URI(string: "file:///" + targetIdentifier.uri.stringValue.split(separator: ":").last! + ".swift")
+        return OutputsItem(target: targetIdentifier, outputPaths: [outputFile])
+      }
+    }
+    buildSystem.targetOutputs = targetOutputResponse
+
+    sk.allowUnexpectedNotification = false
+
+    let targetURI = DocumentURI(string: "target://c:c")
+    let newSettings = IndexVisibility(targets: [BuildTargetIdentifier(uri: targetURI)], includeTargetDependencies: true)
+    let schemeChange: WorkspaceSettingsChange = .sourcekitlsp(SourceKitLSPWorkspaceSettings(indexVisibility: newSettings))
+
+    sk.send(DidChangeConfigurationNotification(settings: schemeChange))
+
+    // Unfortunately there's no callback for scheme change notification from sourcekit server, waiting here for the result
+    let expectation = self.expectation(description: "Waiting for scheme change notification handling")
```

Comment:
We need a real way to test this.  We cannot have tests that fail spuriously in CI because of timing issues.  One idea: you could split up `onIndexVisibilityChange` into a public for testing method and an internal version.  If you pass through a callback you can use that to detect when the change finishes.  It doesn't test the notification code path itself, but it hits the important parts I think.

```swift
/// *Public for testing*
public func _onIndexVisibilityChange(..., completion: @escaping (...)->Void) {
  queue.async { self. onIndexVisibilityChange(..., completion)  }
}

/// Must be called on `queue`.
func onIndexVisibilityChange(settings: IndexVisibility, workspace: Workspace, completion: @escaping (...)->Void) {
...
}
```